### PR TITLE
also print the entries index

### LINKED ISSUE #4524: Hail's type equality check should report a path to the unequal component
Consider:
```
hail.utils.java.FatalError: HailException: optimization changed type!
  before: Matrix{global:+Struct{},col_key:[s],col:Struct{s:String},row_key:[[locus,alleles]],row:Struct{locus:Locus(GRCh38),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{CIEND:Array[Int32],CIPOS:Array[Int32],CS:String,END:Int32,IMPRECISE:Boolean,MC:Array[String],MEINFO:Array[String],MEND:Int32,MLEN:Int32,MSTART:Int32,SVLEN:Array[Int32],SVTYPE:String,TSD:String,AC:Array[Int32],AF:Array[Float64],NS:Int32,AN:Int32,EAS_AF:Array[Float64],EUR_AF:Array[Float64],AFR_AF:Array[Float64],AMR_AF:Array[Float64],SAS_AF:Array[Float64],DP:Int32,AA:String,VT:Array[String],EX_TARGET:Boolean,MULTI_ALLELIC:Boolean,STRAND_FLIP:Boolean,REF_SWITCH:Boolean,DEPRECATED_RSID:Array[String],RSID_REMOVED:Array[String],GRCH37_38_REF_STRING_MATCH:Boolean,NOT_ALL_RSIDS_STRAND_CHANGE_OR_REF_SWITCH:Boolean,GRCH37_POS:Int32,GRCH37_REF:String,ALLELE_TRANSFORM:Boolean,REF_NEW_ALLELE:Boolean,CHROM_CHANGE_BETWEEN_ASSEMBLIES:Array[String]},a_index:Int32,was_split:Boolean,old_locus:Locus(GRCh38),old_alleles:Array[String]},entry:Struct{GT:Call}}
  after:  Matrix{global:+Struct{},col_key:[s],col:Struct{s:String},row_key:[[locus,alleles]],row:Struct{locus:Locus(GRCh38),alleles:Array[String],rsid:String,qual:Float64,filters:Set[String],info:Struct{CIEND:Array[Int32],CIPOS:Array[Int32],CS:String,END:Int32,IMPRECISE:Boolean,MC:Array[String],MEINFO:Array[String],MEND:Int32,MLEN:Int32,MSTART:Int32,SVLEN:Array[Int32],SVTYPE:String,TSD:String,AC:Array[Int32],AF:Array[Float64],NS:Int32,AN:Int32,EAS_AF:Array[Float64],EUR_AF:Array[Float64],AFR_AF:Array[Float64],AMR_AF:Array[Float64],SAS_AF:Array[Float64],DP:Int32,AA:String,VT:Array[String],EX_TARGET:Boolean,MULTI_ALLELIC:Boolean,STRAND_FLIP:Boolean,REF_SWITCH:Boolean,DEPRECATED_RSID:Array[String],RSID_REMOVED:Array[String],GRCH37_38_REF_STRING_MATCH:Boolean,NOT_ALL_RSIDS_STRAND_CHANGE_OR_REF_SWITCH:Boolean,GRCH37_POS:Int32,GRCH37_REF:String,ALLELE_TRANSFORM:Boolean,REF_NEW_ALLELE:Boolean,CHROM_CHANGE_BETWEEN_ASSEMBLIES:Array[String]},a_index:Int32,was_split:Boolean,old_locus:Locus(GRCh38),old_alleles:Array[String]},entry:Struct{GT:Call}}
```
It is not clear how these differ (if at all?). If the type equality checker returns evidence of *why* it thinks the two types are unequal, it is easier for us to debug the type checker and the code should have produced equal types!

### PR DESCRIPTION
**Pull request by danking**
Fixes #4524

Two matrix types can be non-equal if they have different entries locations, so we should include that information in the printed form.

Example:
```
Matrix{global:Struct{},col_key:[col_idx],col:Struct{col_idx:Int32},row_key:[[locus,alleles]],row:Struct{row_idx:Int32,locus:Locus(GRCh37),alleles:Array[String],a_index:Int32,was_split:Boolean,old_locus:Locus(GRCh37),old_alleles:Array[String]},entry:Struct{}entriesIndex:7}
```

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/types/MatrixType.scala [hunk 1]
```diff
@@ -135,6 +135,8 @@ case class MatrixType(
     sb.append(s"entry:$space")
     entryType.pretty(sb, indent, compact)
 
+    sb.append(s"entriesIndex:$space${entriesIdx}")
```
Comment (jigold) [inline hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: Where else is this used? Is it parsable or only for debugging?

Comment (danking) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: @cseed ?

I assume `pretty` is not for parseable things, just for human-readable things.

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: on BaseType:
```
  def parsableString(): String = toPrettyString(0, compact = true)
```

Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: 🤷‍♀️

Comment (danking) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: Uh, is it bad that we don't preserve entries location across a write/read? @tpoterba

Comment (catoverdrive) [reply hail/src/main/scala/is/hail/expr/types/MatrixType.scala]: Rows and entries get split during a write, so they have to get joined back together during a read anyways.

**Comment by danking** [comment]
We'll leave this as a latent issue. I don't want to tease apart what the right answer is right now.

**Pull request closed without merging**